File: backend/app/schemas/delivery_note.py
from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator


# ── Brief models ─────────────────────────────────────────────────────────────

class CustomerBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    first_name: str
    last_name: str
    company: str | None = None


class UserBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


class SalesOrderBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    so_number: str
    title: str


# ── Delivery Note schemas ────────────────────────────────────────────────────

class DeliveryNoteCreate(BaseModel):
    sales_order_id: int
    customer_id: int | None = None
    assigned_to: int | None = None
    delivery_date: date | None = None
    delivery_address: str | None = None
    warehouse: str | None = Field(default=None, max_length=200)
    driver_name: str | None = Field(default=None, max_length=200)
    vehicle_plate: str | None = Field(default=None, max_length=50)
    customer_reference: str | None = Field(default=None, max_length=100)
    terms_conditions: str | None = None
    notes: str | None = None
    internal_notes: str | None = None

    @field_validator("sales_order_id")
    @classmethod
    def sales_order_required(cls, v: int) -> int:
        if not v:
            raise ValueError("sales_order_id is required")
        return v


class DeliveryNoteUpdate(BaseModel):
    customer_id: int | None = None
    assigned_to: int | None = None
    delivery_date: date | None = None
    delivery_address: str | None = None
    warehouse: str | None = None
    driver_name: str | None = None
    vehicle_plate: str | None = None
    customer_reference: str | None = None
    terms_conditions: str | None = None
    notes: str | None = None
    internal_notes: str | None = None


class DeliveryNoteOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    dn_number: str
    status: str
    sales_order: SalesOrderBrief | None = None
    customer: CustomerBrief | None = None
    assigned_user: UserBrief | None = None
    delivery_date: date
    warehouse: str | None = None
    item_count: int = 0
    created_at: datetime
    updated_at: datetime | None = None
    is_active: bool


class DeliveryNoteDetail(DeliveryNoteOut):
    delivery_address: str | None = None
    driver_name: str | None = None
    vehicle_plate: str | None = None
    customer_reference: str | None = None
    notes: str | None = None
    internal_notes: str | None = None
    terms_conditions: str | None = None
    created_by: int | None = None
    updated_by: int | None = None
    items: list["DeliveryNoteItemOut"] = []
    recent_status_history: list["DeliveryNoteStatusHistoryOut"] = []
    available_actions: list[str] = []


class DeliveryNoteListResponse(BaseModel):
    items: list[DeliveryNoteOut]
    total: int
    page: int
    page_size: int
    pages: int


# ── Item schemas ─────────────────────────────────────────────────────────────

class DeliveryNoteItemCreate(BaseModel):
    item_code: str | None = Field(default=None, max_length=100)
    description: str = Field(..., min_length=1, max_length=1000)
    quantity_ordered: float | None = Field(default=None, ge=0)
    quantity_delivered: float = Field(default=0.0, ge=0)
    unit: str = Field(default="unit", max_length=50)
    notes: str | None = Field(default=None, max_length=500)
    sort_order: int = 0


class DeliveryNoteItemBulkItem(DeliveryNoteItemCreate):
    """One row in a bulk `PUT /delivery-notes/{id}/items/bulk` replace payload —
    same Excel-grid-editor Save pattern as Quotation/Sales Order."""
    id: int | None = None


class DeliveryNoteItemOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    delivery_note_id: int
    line_number: int
    item_code: str | None = None
    description: str
    quantity_ordered: float | None = None
    quantity_delivered: float
    unit: str
    notes: str | None = None
    sort_order: int
    created_at: datetime


# ── Status History ───────────────────────────────────────────────────────────

class DeliveryNoteStatusHistoryOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    delivery_note_id: int
    from_status: str | None = None
    to_status: str
    reason: str | None = None
    user: UserBrief | None = None
    changed_at: datetime


# ── Workflow actions ─────────────────────────────────────────────────────────

class WorkflowAction(BaseModel):
    reason: str | None = Field(default=None, max_length=500)


# ── Forward ref rebuild ──────────────────────────────────────────────────────

DeliveryNoteDetail.model_rebuild()

File: frontend/src/types/deliveryNote.ts
export type DeliveryNoteStatus = 'draft' | 'dispatched' | 'delivered' | 'cancelled'

export interface CustomerBrief {
  id: number
  first_name: string
  last_name: string
  company: string | null
}

export interface UserBrief {
  id: number
  username: string
  full_name: string | null
}

export interface SalesOrderBrief {
  id: number
  so_number: string
  title: string
}

export interface DeliveryNoteItemOut {
  id: number
  delivery_note_id: number
  line_number: number
  item_code: string | null
  description: string
  quantity_ordered: number | null
  quantity_delivered: number
  unit: string
  notes: string | null
  sort_order: number
  created_at: string
}

export interface StatusHistoryEntry {
  id: number
  delivery_note_id: number
  from_status: string | null
  to_status: string
  reason: string | null
  user: UserBrief | null
  changed_at: string
}

export interface DeliveryNote {
  id: number
  dn_number: string
  status: DeliveryNoteStatus
  sales_order: SalesOrderBrief | null
  customer: CustomerBrief | null
  assigned_user: UserBrief | null
  delivery_date: string
  warehouse: string | null
  item_count: number
  created_at: string
  updated_at: string | null
  is_active: boolean
}

export interface DeliveryNoteDetail extends DeliveryNote {
  delivery_address: string | null
  driver_name: string | null
  vehicle_plate: string | null
  customer_reference: string | null
  notes: string | null
  internal_notes: string | null
  terms_conditions: string | null
  created_by: number | null
  updated_by: number | null
  items: DeliveryNoteItemOut[]
  recent_status_history: StatusHistoryEntry[]
  available_actions: string[]
}

export interface DeliveryNoteListResponse {
  items: DeliveryNote[]
  total: number
  page: number
  page_size: number
  pages: number
}

export interface DeliveryNoteCreate {
  sales_order_id: number
  customer_id?: number | null
  assigned_to?: number | null
  delivery_date?: string | null
  delivery_address?: string
  warehouse?: string
  driver_name?: string
  vehicle_plate?: string
  customer_reference?: string
  terms_conditions?: string
  notes?: string
  internal_notes?: string
}

export interface DeliveryNoteUpdate {
  customer_id?: number | null
  assigned_to?: number | null
  delivery_date?: string | null
  delivery_address?: string
  warehouse?: string
  driver_name?: string
  vehicle_plate?: string
  customer_reference?: string
  terms_conditions?: string
  notes?: string
  internal_notes?: string
}

export interface DeliveryNoteItemCreate {
  item_code?: string
  description: string
  quantity_ordered?: number | null
  quantity_delivered?: number
  unit?: string
  notes?: string
  sort_order?: number
}

export interface DeliveryNoteItemBulkItem extends DeliveryNoteItemCreate {
  id?: number | null
}

export interface DeliveryNoteListParams {
  q?: string
  status?: DeliveryNoteStatus
  customer_id?: number
  sales_order_id?: number
  assigned_to?: number
  is_active?: boolean
  created_from?: string
  created_to?: string
  page?: number
  page_size?: number
  sort?: 'created_at' | 'updated_at' | 'delivery_date' | 'dn_number' | 'status'
  order?: 'asc' | 'desc'
}
